# Pivot Tables in Pandas 

In pandas, a pivot table is a data processing tool that allows you to reshape and summarize data in a tabular format. It is a way of transforming and restructuring a DataFrame to provide a different perspective on the data. The primary function for creating pivot tables in pandas is `pd.pivot_table()`.

**Key Components of a Pivot Table:**
- **Index:** The rows in the pivot table are defined by one or more columns from the original DataFrame, referred to as the index.
- **Columns:** The columns in the pivot table are defined by values in a specific column (or columns) from the original DataFrame.
- **Values:** The values in the cells of the pivot table are determined by a specified aggregation function applied to a particular column from the original DataFrame.


**Why Pivot Tables are Useful:**

1. **Data Summarization:** Pivot tables allow you to summarize and aggregate data, providing a compact and more understandable representation of complex datasets.

2. **Restructuring Data:** They help in transforming and restructuring data to facilitate easier analysis and visualization.

3. **Multidimensional Analysis:** Pivot tables enable you to analyze data across multiple dimensions, making it easier to identify patterns and trends.

4. **Data Exploration:** They provide a dynamic way to explore data, allowing users to quickly change the arrangement of rows and columns to view different aspects of the dataset.

5. **Reporting:** Pivot tables are commonly used in generating reports, as they can present data in a format that is easy to read and interpret.

In summary, pivot tables in pandas are a powerful tool for reshaping, summarizing, and analyzing data, making it easier for data analysts and scientists to derive insights from complex datasets.

## Objective: <br>
Transform real-world data into a structured dataset.

## Data Source: <br>

Utilizing real-world data from `data.worldbank.org` regarding Water Stress in countries from the mid-20th century to the present. Our objective is to streamline the complex data structure through the application of `pd.pivot_table()`.

In [12]:
'''                     Read the Original Data File
'''
import pandas as pd
data = pd.read_csv('Data_water_stress.csv', index_col=0)
print ('Shape of the Original Dataset : ',data.shape)
data.head(10)

Shape of the Original Dataset :  (778017, 8)


Country  Year  \
0                Luxembourg  1965   
1                    Turkey  1965   
2                   Ecuador  1965   
3                      Fiji  1965   
4                     Congo  1965   
5                    Mexico  1965   
6                   Denmark  1965   
7  United States of America  1965   
8                   Namibia  1965   
9                  Mongolia  1965   

                                           Variable  M49          Unit Symbol  \
0  Long-term average annual precipitation in volume  442  10^9 m3/year          
1   Long-term average annual precipitation in depth  792       mm/year          
2   Long-term average annual precipitation in depth  218       mm/year          
3   Long-term average annual precipitation in depth  242       mm/year          
4                   Groundwater produced internally  178  10^9 m3/year          
5                   Groundwater produced internally  484  10^9 m3/year          
6                   Groundwater produced internally  208  10^9 m3/year          
7                   Groundwater produced internally  840  10^9 m3/year          
8                   Groundwater produced internally  516  10^9 m3/year          
9                   Groundwater produced internally  496  10^9 m3/year          

                                  Symbol_Description     Value  
0  Official value, data come from one national so...     2.415  
1  Official value, data come from one national so...   593.000  
2  Official value, data come from one national so...  2274.000  
3  Official value, data come from one national so...  2592.000  
4  Official value, data come from one national so...   122.000  
5  Official value, data come from one national so...   150.000  
6  Official value, data come from one national so...     4.300  
7  Official value, data come from one national so...  1383.000  
8  Official value, data come from one national so...     2.100  
9  Official value, data come from one national so...     6.100

#### Discussion:

Observing the original dataset, it encompasses 8 attributes: `Country, Year, Variable, M49, Unit, Symbol, Symbol_Description, Value`. Each row signifies the `Value` of a unique `Variable` for each `Year` in a specific `Country`. The dataset's intricacy lies in consolidating all Water Stress-defining Variables into a single column.

The dataset features over 180 unique variables crucial for comprehending global Water conditions. However, organizing this dataset is a formidable task. The subsequent discussion delves into employing a Pivot table to address this complexity.

In [17]:
'''                     Uniques attributes in the column "Variable"
'''

print ('The Number of Unique attributes in the column "Variable" is : ', len(data['Variable'].unique()),'\n','\n')
print ('The Unique attributes in the column "Variable" are : \n', data['Variable'].unique())

The Number of Unique attributes in the column "Variable" is :  186 
 

The Unique attributes in the column "Variable" are : 
 ['Long-term average annual precipitation in volume'
 'Long-term average annual precipitation in depth'
 'Groundwater produced internally' 'Surface water produced internally'
 'Overlap between surface water and groundwater'
 'Total internal renewable water resources (IRWR)'
 'Surface water: entering the country (total)'
 'Surface water: inflow not submitted to treaties'
 'Surface water: inflow submitted to treaties'
 'Surface water: inflow secured through treaties'
 'Surface water: total flow of border rivers'
 'Surface water: accounted flow of border rivers'
 'Surface water: leaving the country to other countries (total)'
 'Surface water: outflow to other countries not submitted to treaties'
 'Surface water: outflow to other countries submitted to treaties'
 'Surface water: outflow to other countries secured through treaties'
 'Groundwater: entering the country 

#### Opinion:

The utilization of a Pivot table for restructuring such intricate datasets is a prudent approach. It not only simplifies the representation of data but also enables dynamic exploration and analysis. Pivot tables are invaluable tools for data scientists, providing an effective means to glean insights from multifaceted datasets.

## Task - Read-Select-Execute-Add <br>

The task is divided into four distinct steps:

**Read:** Demonstrates the process of reading the original CSV file.

**Select:** Guides you through the selection of specific variables of interest from the pool of 186 variables.

**Execute:** Involves the application of `pd.pivot_table` to the chosen variables.

**Add:** Introduces the addition of a new column named `Class_of_waterstress` based on Water Stress values.

In [19]:
from pylab import *
import pandas as pd
import seaborn as sns
sys.setrecursionlimit(10000)

def extract_data(data):
    data_waterstress = data[(data.Variable == 'SDG 6.4.2. Water Stress') | (data.Variable == 'Total population')
                           | (data.Variable == 'Long-term average annual precipitation in volume')
                           | (data.Variable == 'Total freshwater withdrawal')
                           | (data.Variable == 'Agricultural water withdrawal')
                           | (data.Variable == 'Industrial water withdrawal')]
    df_out = pd.pivot_table(data_waterstress, index=['Year','Country'],columns="Variable", values="Value") #data_waterstress
    return (df_out)

def add_classofwaterstress(df):

    # create a list of our conditions
    conditions = [
        (df['SDG 6.4.2. Water Stress'] > 0.) & (df['SDG 6.4.2. Water Stress'] <= 25.),
        (df['SDG 6.4.2. Water Stress'] > 25.) & (df['SDG 6.4.2. Water Stress'] <= 50.),
        (df['SDG 6.4.2. Water Stress'] > 50.) & (df['SDG 6.4.2. Water Stress'] <= 75.),
        (df['SDG 6.4.2. Water Stress'] > 75.) & (df['SDG 6.4.2. Water Stress'] <= 100.),
        (df['SDG 6.4.2. Water Stress'] > 100.)
        ]

    # create a list of the values we want to assign for each condition
    values = ['0-25% (No-stress)', '25-50% (Low)', '50-75% (Medium)', '75-100% (High)', ' >100% (Critical)']

    # create a new column and use np.select to assign values to it using our lists as arguments
    df['Class_of_waterstress'] = select(conditions, values)
    return(df)

def save_results(results):
    writer = pd.ExcelWriter('pivot_table.xlsx', engine='xlsxwriter')
    results.to_excel(writer,sheet_name='dataframe')
    writer.save()

def uniques(df):
    group = df.groupby('Variable')
    df2 = group.apply(lambda x: x['Unit'].unique())
    writer = pd.ExcelWriter('Unique_variables&Units.xlsx', engine='xlsxwriter')
    df2.to_excel(writer,sheet_name='dataframe')
    writer.save()


if __name__ == "__main__":
    
    data = pd.read_csv('Data_water_stress.csv', index_col=0)
    results = extract_data(data)
    final_form = add_classofwaterstress(results)
    save_results(final_form) 
    uniques(data)
    
    

C:\Users\AK traders\AppData\Local\Temp\ipykernel_1564\940136455.py:36: FutureWarning: save is not part of the public API, usage can give unexpected results and will be removed in a future version
  writer.save()
C:\Users\AK traders\AppData\Local\Temp\ipykernel_1564\940136455.py:43: FutureWarning: save is not part of the public API, usage can give unexpected results and will be removed in a future version
  writer.save()


In [22]:
'''                      Outcome Results
'''
print ('Shape of the Outcome Dataframe : ', final_form.shape,'\n\n')
final_form

Shape of the Outcome Dataframe :  (10832, 7) 




Variable                                 Agricultural water withdrawal  \
Year Country                                                             
1961 Afghanistan                                                   NaN   
     Albania                                                       NaN   
     Algeria                                                       NaN   
     Andorra                                                       NaN   
     Angola                                                        NaN   
...                                                                ...   
2019 Venezuela (Bolivarian Republic of)                       16.71000   
     Viet Nam                                                 77.75000   
     Yemen                                                     3.23500   
     Zambia                                                    1.15200   
     Zimbabwe                                                  3.04107   

Variable                                 Industrial water withdrawal  \
Year Country                                                           
1961 Afghanistan                                                 NaN   
     Albania                                                     NaN   
     Algeria                                                     NaN   
     Andorra                                                     NaN   
     Angola                                                      NaN   
...                                                              ...   
2019 Venezuela (Bolivarian Republic of)                      0.79330   
     Viet Nam                                                3.07400   
     Yemen                                                   0.06500   
     Zambia                                                  0.13000   
     Zimbabwe                                                0.08135   

Variable                                 Long-term average annual precipitation in volume  \
Year Country                                                                                
1961 Afghanistan                                                                213.48522   
     Albania                                                                     42.69375   
     Algeria                                                                    211.97486   
     Andorra                                                                          NaN   
     Angola                                                                    1259.16700   
...                                                                                   ...   
2019 Venezuela (Bolivarian Republic of)                                        1864.23020   
     Viet Nam                                                                   603.16983   
     Yemen                                                                       88.17099   
     Zambia                                                                     767.66220   
     Zimbabwe                                                                   256.72932   

Variable                                 SDG 6.4.2. Water Stress  \
Year Country                                                       
1961 Afghanistan                                             NaN   
     Albania                                                 NaN   
     Algeria                                                 NaN   
     Andorra                                                 NaN   
     Angola                                                  NaN   
...                                                          ...   
2019 Venezuela (Bolivarian Republic of)                 7.540367   
     Viet Nam                                          18.130315   
     Yemen                                            169.761905   
     Zambia                                             2.835498   
     Zimbabwe                                          35.405370   

Varia

## Conlcusion

In this brief project, I curated real-world data on Water Stress and employed `pd.pivot_table()` to restructure the dataset into a more organized format. The resultant dataset now encapsulates crucial variables as attributes, delineating Water Stress for each country in every year.

**Opinion:**
Using `pd.pivot_table()` for such tasks aligns with best practices in data manipulation. It's an efficient method for transforming complex datasets into a more understandable and analyzable structure. The inclusion of a new column, `Class_of_waterstress`, enhances the dataset's utility by providing an additional layer of information.

## Future Work

This transformed dataset holds potential for research purposes. While the current project focused on structural transformation, the dataset could serve as a valuable foundation for future endeavors. Machine Learning applications could leverage this original dataset to delve deeper into understanding the primary causes and parameters influencing water stress. Although beyond the scope of this project, implementation in the near future could unlock further insights.